File: examples/javascript/json-transform/src/handler.ts
import type { Handler, JsonObject, JsonValue } from "nanofaas-function-sdk";

function isJsonObject(value: unknown): value is JsonObject {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function remapObject(data: JsonObject, fieldMap: Record<string, string>): JsonObject {
    const output: JsonObject = {};
    for (const [key, value] of Object.entries(data)) {
        output[fieldMap[key] ?? key] = value as JsonValue;
    }
    return output;
}

export const handleJsonTransform: Handler = async (ctx, req) => {
    ctx.logger.info("processing json transform");

    if (!isJsonObject(req.input)) {
        return { error: "Input must be a JSON object" };
    }

    const data = req.input.data;
    if (!isJsonObject(data)) {
        return { error: "Field 'data' must be a JSON object" };
    }

    const rawFieldMap = req.input.fieldMap;
    const fieldMap: Record<string, string> = isJsonObject(rawFieldMap)
        ? Object.fromEntries(
            Object.entries(rawFieldMap)
                .filter((entry): entry is [string, string] => typeof entry[1] === "string"),
        )
        : {};

    return remapObject(data, fieldMap);
};

File: functions/python/json-transform/handler.py
from nanofaas.sdk import nanofaas_function, context
from statistics import mean

logger = context.get_logger(__name__)

@nanofaas_function
def handle(input_data):
    logger.info(f"Processing json-transform for execution {context.get_execution_id()}")

    if not isinstance(input_data, dict):
        return {"error": "Input must be a JSON object"}
        
    data = input_data.get("data")
    group_by = input_data.get("groupBy")
    operation = input_data.get("operation", "count")
    value_field = input_data.get("valueField")

    if data is None or group_by is None:
        return {"error": "Fields 'data' and 'groupBy' are required"}

    if operation != "count" and not value_field:
        return {"error": f"Field 'valueField' is required for operation: {operation}"}

    grouped = {}
    for item in data:
        key = str(item.get(group_by, "null"))
        if key not in grouped:
            grouped[key] = []
        grouped[key].append(item)

    result_groups = {}
    for key, items in grouped.items():
        if operation == "count":
            val = len(items)
        else:
            values = [i.get(value_field) for i in items if i.get(value_field) is not None]
            if not values:
                val = 0
            elif operation == "sum":
                val = sum(values)
            elif operation == "avg":
                val = mean(values)
            elif operation == "min":
                val = min(values)
            elif operation == "max":
                val = max(values)
            else:
                val = f"unknown operation: {operation}"
            
        result_groups[key] = val

    return {
        "groupBy": group_by,
        "operation": operation,
        "groups": result_groups
    }
